test/exec/test_task.cpp

```diff
@@ -0,0 +1,172 @@
+    co_await (
+      test_stickiness_for_two_single_thread_contexts_nested(scheduler1, scheduler2, id1, id2)
+      | then([&] { CHECK(std::this_thread::get_id() == id2); }));
+  }
+
+  task<void> test_stickiness_for_two_single_thread_contexts(
+    auto scheduler1,
+    auto scheduler2,
+    auto id1,
+    auto id2) {
+    co_await complete_inline(scheduler1);
+    CHECK(std::this_thread::get_id() == id1);
+    co_await test_stickiness_for_two_single_thread_contexts_(scheduler1, scheduler2, id1, id2);
+    CHECK(std::this_thread::get_id() == id1);
+  }
+
+  task<void> test_stickiness_for_two_single_thread_contexts_with_sender(
+    auto scheduler1,
+    auto scheduler2,
+    auto id1,
+    auto id2) {
+    co_await complete_inline(scheduler1);
+    CHECK(std::this_thread::get_id() == id1);
+    co_await test_stickiness_for_two_single_thread_contexts_with_sender_(
+      scheduler1, scheduler2, id1, id2);
+    CHECK(std::this_thread::get_id() == id1);
+  }
+}
+
+TEST_CASE("Test stickiness with two single threads", "[types][sticky][task]") {
+  single_thread_context context1;
+  single_thread_context context2;
+  scheduler auto scheduler1 = context1.get_scheduler();
+  scheduler auto scheduler2 = context2.get_scheduler();
+  auto id1 = context1.get_thread_id();
+  auto id2 = context2.get_thread_id();
+  auto t = test_stickiness_for_two_single_thread_contexts(scheduler1, scheduler2, id1, id2);
+  sync_wait(std::move(t));
+}
+
+TEST_CASE("Test stickiness with two single threads with sender", "[types][sticky][task]") {
+  single_thread_context context1;
+  single_thread_context context2;
+  scheduler auto scheduler1 = context1.get_scheduler();
+  scheduler auto scheduler2 = context2.get_scheduler();
+  auto id1 = context1.get_thread_id();
+  auto id2 = context2.get_thread_id();
+  auto t = test_stickiness_for_two_single_thread_contexts_with_sender(
+    scheduler1, scheduler2, id1, id2);
+  sync_wait(std::move(t));
+}
+
+TEST_CASE("Use two inline schedulers", "[types][sticky][task]") {
+  scheduler auto scheduler1 = exec::inline_scheduler{};
+  scheduler auto scheduler2 = exec::inline_scheduler{};
+  auto id1 = std::this_thread::get_id();
+  auto id2 = std::this_thread::get_id();
+  auto t = test_stickiness_for_two_single_thread_contexts(scheduler1, scheduler2, id1, id2);
+  sync_wait(std::move(t));
+}
+
+namespace {
+  task<void> test_old_behaviour_nested(
+    scheduler auto sched1,
+    scheduler auto sched2,
+    auto id_main_thread,
+    auto id1,
+    auto id2) {
+    CHECK(std::this_thread::get_id() == id2);
+    co_await schedule(sched1);
+    CHECK(std::this_thread::get_id() == id1);
+  }
+
+  task<void> test_old_behaviour(
+    scheduler auto sched1,
+    scheduler auto sched2,
+    auto id_main_thread,
+    auto id1,
+    auto id2) {
+    CHECK(std::this_thread::get_id() == id_main_thread);
+    co_await schedule(sched1);
+    CHECK(std::this_thread::get_id() == id1); // We changed from main thread to context1
```

Comment:
This should be on the main thread. The rule is: for any sender not wrapped in a complete_inline, transfer back to what was the "current" scheduler before the co_await.

I'd also like any use of complete_inline to cause an at_coroutine_exit action to be posted that transfers back to the starting scheduler.

Reply:
I need to check since I'm on phone too. But I think the current scheduler is inline scheduler and that's why, it behaves non sticky

Reply:
Ok, I changed the implementation to use `stdexec::get_scheduler` and `stdexec::__scheduler_provider`. Now The test performs as *you* expect it will be rescheduled on the main thread. This is because the task takes scheduler from the `sync_wait`'s run_loop. This is so exciting!

Reply:
> I'd also like any use of complete_inline to cause an at_coroutine_exit action to be posted that transfers back to the starting scheduler.

You are right. The task is currently always rescheduled. It should be better to do that only if we used complete inline. I was slow at that one and will change that asap 😍